### PR TITLE
Fix consent request for attributes that don't have values.

### PR DESCRIPTION
**Pull request by Wathsara**
#### Purpose

Previously consent has been asked for attributes that don't have values in the profile. With this fix the consent will be asked for the user attributes that have values.

Depends on - Depends on https://github.com/wso2/product-is/pull/12158

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -2253,17 +2260,42 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+            if (log.isDebugEnabled()) {
+                log.debug(errorMsg, e);
+            }
+            log.error(errorMsg);
+        } catch (UserStoreException e) {
+            String errorMsg = "Error While getting the User Store or fetching claim values. ";
+            if (log.isDebugEnabled()) {
+                log.debug(errorMsg, e);
+            }
+            log.error(errorMsg);
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2284]: Do we need to add debug and error logs both?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 2]
```diff
@@ -2253,17 +2260,42 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+            AbstractUserStoreManager userStore = (AbstractUserStoreManager) realm.getUserStoreManager();
+            // Get the claims which have values from external claims of requested scope and essential claims.
+            localClaimsOfOidcClaimsWithValues = userStore
+                    .getUserClaimValuesWithID(user.getUserId(), localClaimsOfOidcClaims.toArray(new String[0]), null);
+            // Get the claims which have values from external claims relevant to all essential requested claims.
+            localClaimsOfEssentialClaimsWithValues = userStore
+                    .getUserClaimValuesWithID(user.getUserId(), localClaimsOfEssentialClaims.toArray(new String[0]),
+                            null);
+        } catch (IdentityException e) {
+            String errorMsg = "Error While getting the User Realm. ";
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2274]: Shall we add the username and the tenant domain to the error log

**Comment by GANGANI** [comment]
@Wathsara Shall we fix this in post consent handler in the framework as well? If you have already fixed it, please link that PR to the description.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 3]
```diff
@@ -2177,6 +2185,42 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+                localClaimUris.add(externalClaim.getMappedLocalClaim());
+            }
+        }
+        for (ClaimMetaData claimMetaData : requestedClaims) {
+            if (localClaimUris.contains(claimMetaData.getClaimUri())) {
+                filteredRequestClaims.add(claimMetaData);
+            }
+        }
+        return filteredRequestClaims;
+
```
Comment (mevan-karu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2221]:
```suggestion
```

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 4]
```diff
@@ -2177,6 +2185,42 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+     * @param requestedClaims           List of requested claims metadata.
+     * @param userAttributes            Authenticated users' attributes.
+     * @param spTenantDomain            Tenant domain.
+     * @return                          Filtered claims with user attributes.
+     * @throws ClaimMetadataException   If an error occurred while getting claim mappings.
+     */
+    private List<ClaimMetaData> removeClaimsNotInUserAttributes(List<ClaimMetaData> requestedClaims,
+                                                                Map<ClaimMapping, String> userAttributes,
+                                                                String spTenantDomain) throws ClaimMetadataException {
+
```
Comment (mevan-karu) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2200]: Check null/empty for `requestedClaims` and `userAttributes`. If either of them is null/empty return empty list of `ClaimMetaData`

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 5]
```diff
@@ -2177,6 +2185,42 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
         return additionalQueryParam;
     }
 
+    /**
+     * Filter out the requested claims with the user attributes.
+     *
+     * @param requestedClaims           List of requested claims metadata.
+     * @param userAttributes            Authenticated users' attributes.
+     * @param spTenantDomain            Tenant domain.
+     * @return                          Filtered claims with user attributes.
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2194]: Check formatting

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 6]
```diff
@@ -2177,6 +2185,42 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+     * @throws ClaimMetadataException   If an error occurred while getting claim mappings.
+     */
+    private List<ClaimMetaData> removeClaimsNotInUserAttributes(List<ClaimMetaData> requestedClaims,
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2197]:
```suggestion
    private List<ClaimMetaData> removeConsentRequestedNullUserAttributes(List<ClaimMetaData> requestedClaims,
```
WDYT?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 4 above)
Comment (Wathsara) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2200]: Fixed

**Review by mevan-karu — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 7]
```diff
@@ -2177,6 +2184,45 @@ private String handlePreConsent(OAuth2Parameters oauth2Params, AuthenticatedUser
...
+                localClaims.add(attribute.getKey().getLocalClaim().getClaimUri());
+            }
+            if (CollectionUtils.isNotEmpty(localClaims)) {
+                Set<ExternalClaim> externalClaimSetOfOidcClaims = ClaimMetadataHandler.getInstance()
+                        .getMappingsFromOtherDialectToCarbon(OIDC_DIALECT, new HashSet<String>(localClaims),
+                                spTenantDomain);
+                for (ExternalClaim externalClaim : externalClaimSetOfOidcClaims) {
+                    localClaimUris.add(externalClaim.getMappedLocalClaim());
+                }
+            }
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2216]: Why is this needed? What is the difference between localClaims and localClaimUris?

**Review by IsuraD — APPROVED** [review]
